File: iOSCode/MyLocationsOC/MapViewController.m
//
//  MapViewController.m
//  MyLocations
//
//  Created by ZhangZehua on 2/22/17.
//  Copyright © 2017 ZhangZehua. All rights reserved.
//

#import "Location+CoreDataClass.h"
#import "MapViewController.h"
#import "GeneralFunctions.h"
#import "LocationDetailsViewController.h"

@interface MapViewController () <MKMapViewDelegate, UINavigationBarDelegate>

@property (nonatomic, weak) IBOutlet MKMapView* mapView;

@end

@implementation MapViewController {
  NSArray* _locations;
}

- (void)viewDidLoad {
    [super viewDidLoad];
  [self updateLocations];
  
  if ([_locations count] > 0) {
    [self showLocations];
  }
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
  if ((self = [super initWithCoder:aDecoder])) {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(contextDidChange:) name:NSManagedObjectContextObjectsDidChangeNotification object:self.managedObjectContext];
  }
  return self;
}

- (void)contextDidChange:(NSNotification*)notification {
  if ([self isViewLoaded]) {
    [self updateLocations];
  }
}

- (void)dealloc {
  [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (MKCoordinateRegion)regionForAnnotations:(NSArray*)annotations {
  MKCoordinateRegion region;
  if ([annotations count] == 0) {
    region = MKCoordinateRegionMakeWithDistance(self.mapView.userLocation.coordinate, 1000, 1000);
  } else if ([annotations count] == 1) {
    id <MKAnnotation> annotation = [annotations lastObject];
    region = MKCoordinateRegionMakeWithDistance(annotation.coordinate, 1000, 1000);
  } else {
    CLLocationCoordinate2D topLeftCoordinate;
    topLeftCoordinate.latitude = -90;
    topLeftCoordinate.longitude = 180;
    
    CLLocationCoordinate2D bottomRightCoordinate;
    bottomRightCoordinate.latitude = 90;
    bottomRightCoordinate.longitude = -180;
    
    for (id <MKAnnotation> annotation in annotations) {
      topLeftCoordinate.latitude = fmax(topLeftCoordinate.latitude, annotation.coordinate.latitude);
      topLeftCoordinate.longitude = fmin(topLeftCoordinate.longitude, annotation.coordinate.longitude);
      bottomRightCoordinate.latitude = fmin(bottomRightCoordinate.latitude, annotation.coordinate.latitude);
      bottomRightCoordinate.longitude = fmax(bottomRightCoordinate.longitude, annotation.coordinate.longitude);
    }
    const double extraSpace = 1.1;
    region.center.latitude = topLeftCoordinate.latitude - (topLeftCoordinate.latitude - bottomRightCoordinate.latitude) / 2.0;
    region.center.longitude = topLeftCoordinate.longitude - (topLeftCoordinate.longitude - bottomRightCoordinate.longitude) / 2.0;
    region.span.latitudeDelta = fabs(topLeftCoordinate.latitude - bottomRightCoordinate.latitude) * extraSpace;
    region.span.longitudeDelta = fabs(topLeftCoordinate.longitude - bottomRightCoordinate.longitude) * extraSpace;
  }
  return [self.mapView regionThatFits:region];
}

- (void)showLocationDetails:(UIButton*)button {
  [self performSegueWithIdentifier:@"EditLocation" sender:button];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
  if ([segue.identifier isEqualToString:@"EditLocation"]) {
    UINavigationController* nvc = segue.destinationViewController;
    LocationDetailsViewController* locationDetailVC = (LocationDetailsViewController*)nvc.topViewController;
    locationDetailVC.managedObjectContext = self.managedObjectContext;
    UIButton* button = (UIButton*)sender;
    Location* location = _locations[button.tag];
    locationDetailVC.locationToEdit = location;
  }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)showUser {
  MKCoordinateRegion region = MKCoordinateRegionMakeWithDistance(self.mapView.userLocation.coordinate, 1000, 1000);
  [self.mapView setRegion:[self.mapView regionThatFits:region] animated:YES];
}

- (IBAction)showLocations {
  MKCoordinateRegion region = [self regionForAnnotations:_locations];
  [self.mapView setRegion:region animated:YES];
}

- (void)updateLocations {
  NSEntityDescription* entity = [NSEntityDescription entityForName:@"Location" inManagedObjectContext:self.managedObjectContext];
  NSFetchRequest* fetchRequest = [[NSFetchRequest alloc] init];
  [fetchRequest setEntity:entity];
  
  NSError* error;
  NSArray* foundObjects = [self.managedObjectContext executeFetchRequest:fetchRequest error:&error];
  if (foundObjects == nil) {
    [GeneralFunctions fatalCoreDataError:error];
    return;
  }
  
  if (_locations != nil) {
    [self.mapView removeAnnotations:_locations];
  }
  
  _locations = foundObjects;
  [self.mapView addAnnotation:_locations];
}

#pragma mark - MKMapViewDelegate

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id<MKAnnotation>)annotation {
  if ([annotation isKindOfClass:[Location class]]) {
    static NSString* identifier = @"Location";
    MKPinAnnotationView* annotationView = (MKPinAnnotationView*)[self.mapView dequeueReusableAnnotationViewWithIdentifier:identifier];
    if (annotationView == nil) {
      annotationView = [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:identifier];
      annotationView.enabled = YES;
      annotationView.canShowCallout = YES;
      annotationView.animatesDrop = NO;
      annotationView.pinTintColor = [UIColor greenColor];
      
      UIButton* rightbutton = [UIButton buttonWithType:UIButtonTypeDetailDisclosure];
      [rightbutton addTarget:self action:@selector(showLocationDetails:) forControlEvents:UIControlEventTouchUpInside];
      annotationView.rightCalloutAccessoryView = rightbutton;
    } else {
      annotationView.annotation = annotation;
    }
    UIButton* button = (UIButton*)annotationView.rightCalloutAccessoryView;
    button.tag = [_locations indexOfObject:(Location*)annotation];
    return annotationView;
  }
  return nil;
}

#pragma mark - UINavigationBarDelegate

- (UIBarPosition)positionForBar:(id<UIBarPositioning>)bar {
  return UIBarPositionTopAttached;
}

@end

File: iOSCode/MyLocations/MapViewController.swift
//
//  MapViewController.swift
//  MyLocations
//
//  Created by ZhangZehua on 12/9/15.
//  Copyright © 2015 ZhangZehua. All rights reserved.
//

import UIKit
import MapKit
import CoreData

class MapViewController: UIViewController {
    
    //MARK:- IBOutlet
    @IBOutlet weak var mapView: MKMapView!
    
    //MARK:- property
    var managedObjectContext: NSManagedObjectContext! {
        didSet {
            NSNotificationCenter.defaultCenter().addObserverForName(
                NSManagedObjectContextObjectsDidChangeNotification,
                object: managedObjectContext,
                queue: NSOperationQueue.mainQueue()) {
                    notification in
                    if self.isViewLoaded() {
                        self.updateLocations()
                    }
//                    if let dictionary = notification.userInfo {
//                        print(dictionary["inserted"])
//                        print(dictionary["deleted"])
//                        print(dictionary["updated"])
//                    }
            }
        }
    }
    var locations = [Location]()
    
    //MARK:- IBAction
    @IBAction func showUser() {
        let region = MKCoordinateRegionMakeWithDistance(mapView.userLocation.coordinate, 1000, 1000)
        mapView.setRegion(mapView.regionThatFits(region), animated: true)
    }
    
    @IBAction func showLocations() {
        let region = regionForAnnotations(locations)
        mapView.setRegion(region, animated: true)
    }
    
    //MARK:- view
    override func viewDidLoad() {
        super.viewDidLoad()

        updateLocations()
        
       if !locations.isEmpty {
            showLocations()
       }
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "EditLocation" {
            let navigationController = segue.destinationViewController as! UINavigationController
            let controller = navigationController.topViewController as! LocationDetailsViewController
            
            controller.managedObjectContext = managedObjectContext
            
            let button = sender as! UIButton
            let location = locations[button.tag]
            controller.locationToEdit = location
        }
    }
    
    //MARK:- misc
    func updateLocations() {
        let entity = NSEntityDescription.entityForName("Location", inManagedObjectContext: managedObjectContext)
        let fetchRequest = NSFetchRequest()
        fetchRequest.entity = entity
        
        var foundObject: [AnyObject];
        
        do {
            foundObject = try managedObjectContext.executeFetchRequest(fetchRequest)
            
        } catch let error as NSError {
            fatalCoreDataError(error)
            return
        }
        
        mapView.removeAnnotations(locations)
        locations = foundObject as! [Location]
        mapView.addAnnotations(locations)
    }
    
    func regionForAnnotations(annotations: [MKAnnotation]) -> MKCoordinateRegion {
        var region: MKCoordinateRegion
        
        switch annotations.count {
        case 0:
            region = MKCoordinateRegionMakeWithDistance(mapView.userLocation.coordinate, 1000, 1000)
            
        case 1:
            let annotation = annotations[annotations.count - 1]
            region = MKCoordinateRegionMakeWithDistance(annotation.coordinate, 1000, 1000)
        default:
            var topLeftCoord = CLLocationCoordinate2D(latitude: -90, longitude: 180)
            var bottomRightCoord = CLLocationCoordinate2D(latitude: 90, longitude: -180)
            
            for annotation in annotations {
                topLeftCoord.latitude = max(topLeftCoord.latitude, annotation.coordinate.latitude)
                topLeftCoord.longitude = min(topLeftCoord.longitude, annotation.coordinate.longitude)
                bottomRightCoord.latitude = min(bottomRightCoord.latitude, annotation.coordinate.latitude)
                bottomRightCoord.longitude = max(bottomRightCoord.longitude, annotation.coordinate.longitude)
            }
            
            let center = CLLocationCoordinate2D(
                latitude: (topLeftCoord.latitude + bottomRightCoord.latitude) / 2,
                longitude: (topLeftCoord.longitude + bottomRightCoord.longitude) / 2)
            
            let extraSpace = 1.1
            let span = MKCoordinateSpan(
                latitudeDelta: abs(topLeftCoord.latitude - bottomRightCoord.latitude) * extraSpace,
                longitudeDelta: abs(topLeftCoord.longitude - bottomRightCoord.longitude) * extraSpace)
            
            region = MKCoordinateRegion(center: center, span: span)
        }
        return mapView.regionThatFits(region)
    }
    
    func showLocationDetails(sender: UIButton) {
        performSegueWithIdentifier("EditLocation", sender: sender)
    }
}

extension MapViewController: MKMapViewDelegate {
    func mapView(mapView: MKMapView, viewForAnnotation annotation: MKAnnotation) -> MKAnnotationView? {
        if annotation is Location {
            let identifier = "Location"
            var annotationView = mapView.dequeueReusableAnnotationViewWithIdentifier(identifier) as! MKPinAnnotationView!
            
            if annotationView == nil {
                annotationView = MKPinAnnotationView(annotation: annotation, reuseIdentifier: identifier)
                
                annotationView.enabled = true
                annotationView.canShowCallout = true
                annotationView.animatesDrop = false
                annotationView.pinTintColor = UIColor.greenColor()
                annotationView.tintColor = UIColor(white: 0, alpha: 0.5)
                
                let rightButton = UIButton(type: .DetailDisclosure)
                rightButton.addTarget(
                    self,
                    action: Selector("showLocationDetails:"),
                    forControlEvents: .TouchUpInside)
                annotationView.rightCalloutAccessoryView = rightButton
            } else {
                annotationView.annotation = annotation
            }
            
            let button = annotationView.rightCalloutAccessoryView as! UIButton
            if let index = locations.indexOf(annotation as! Location) {
                button.tag = index
            }
            
            return annotationView
        }
        return nil
    }
}

extension MapViewController: UINavigationBarDelegate {
    func positionForBar(bar: UIBarPositioning) -> UIBarPosition {
        return .TopAttached
    }
}
